### DISCUSSION TITLE
What happens during training?

### DISCUSSION THREAD

**Discussion by erezinman** [awslabs/gluonts#1621] [Q&A] [answered]
Hi,
I have a few questions regarding the training process, and how can it be altered if needed:
1. I'm trying to understand what happens during training. Say I want to train a model over a time-series, then what is the model fed during training? Does it sample multiple FCTs (Forecast Creation Times. i.e. time-windows) from each sequence, or does it use the forking-sequences scheme described in the MQRNN article? 
2. How does a batch get filled? Is it filled from multiple sequences exclusively (e.g. from 32 different sequences in the case of batch-size=32), or can there be multiple windows from the same sequence in a batch?
3. Is there a way to use cross-validation techniques for time-series like the one in the [SKLearn](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.TimeSeriesSplit.html) package.
4. Is there a way to limit the forecast-window locations? For instance, I might want the model to take into account a past windows, but to not attempt forecast of the beginning of that window because it cannot be predicted, but should be used for later predictions?
5. Finally, is there a way to perform a multi-scale prediction for non-autoregressive models (say predict 1 hour ahead, 5 hours ahead, 1 day ahead and 1 week ahead)?

Thank you sincerely

**Comment by mbohlkeschneider** [comment, marked as answer]
Hi @erezinman,

Thank you for your questions. 

The short answer is, "it depends" :D. Not everything is true for every model and it depends on it is set up exactly in the transformation chain. Do you have questions about a specific model? Then I can give more accurate answers. I'll try to do my best answering the questions in more general terms: 

>  I'm trying to understand what happens during training. Say I want to train a model over a time-series, then what is the model fed during training? Does it sample multiple FCTs (Forecast Creation Times. i.e. time-windows) from each sequence, or does it use the forking-sequences scheme described in the MQRNN article?

This depends on the model. MQRNN/CNN use the forking scheme, other models like DeepAR sample time points which each batch. But models can be setup differently. If you are interested in a specific model, I can answer more accurately. 

> How does a batch get filled? Is it filled from multiple sequences exclusively (e.g. from 32 different sequences in the case of batch-size=32), or can there be multiple windows from the same sequence in a batch?

Again, this depends on the specific sampler that is setup. Many models use the `ExpectedNumInstanceSampler` where on average each time series is drawn ones per pass over the dataset. 

> Is there a way to use cross-validation techniques for time-series like the one in the SKLearn package.

You can code that up yourself, but GluonTS has no mechanism. However, you can pass a validation set to calculate validation scores during training. 

> Is there a way to limit the forecast-window locations? For instance, I might want the model to take into account a past windows, but to not attempt forecast of the beginning of that window because it cannot be predicted, but should be used for later predictions?

Sorry, I did not understand this. 

> Finally, is there a way to perform a multi-scale prediction for non-autoregressive models (say predict 1 hour ahead, 5 hours ahead, 1 day ahead and 1 week ahead)?

As far as I know, no.

Hope that helps!
